File: ICPC/The 2020 ICPC Asia Yokohama Regional Contest/B.cpp
#include <bits/stdc++.h>
using namespace std;
bool cmp(pair<string, long long> a, pair<string, long long> b)
{
    return a.second > b.second;
}
int main()
{
    long long n;
    cin >> n;
    map<string, long long> d4, d2;
    for (int i = 1; i <= n; i++)
    {
        string s;
        cin >> s;
        string t1 = s.substr(2, 5);
        string t2 = s.substr(4, 5);
        // cout<<t1<<" "<<t2<<endl;
        d4[t1]++;
        d2[t2]++;
    }
    vector<pair<string, long long>> v1, v2;
    long long maxn = 0;
    for (auto x : d4)
        v1.push_back({x.first, x.second});
    for (auto x : d2)
        v2.push_back({x.first, x.second});

    sort(v2.begin(), v2.end(), cmp);
    // for (auto x:v1) cout<<x.first<<" "<<x.second<<endl;
    // for (auto x:v2) cout<<x.first<<" "<<x.second<<endl;

    vector<pair<string, long long>> temp;
    for (auto x : v2)
    {
        temp.push_back(x);
        if (temp.size() == 4)
            break;
    }
    if (temp.size() == 1)
        maxn = max({maxn, (long long)temp[0].second * 500, (long long)300000});
    else if (temp.size() == 4)
        maxn = max({maxn, (long long)500 * (temp[0].second + temp[1].second + temp[2].second) + 300000});
    else
    {
        long long d = 0;
        for (int i = 0; i < temp.size(); i++)
        {
            maxn = max(maxn, (long long)d * 500 + 300000);
            d = (long long)d + temp[i].second;
            maxn = max(maxn, (long long)d * 500);
        }
    }
    // cout<<maxn<<"\n";
    for (auto y : v1)
    {
        vector<pair<string, long long>> temp;
        maxn = max(maxn, (long long)y.second * 4000 + 300000 * (n != y.second));
        long long val = 0;
        for (auto x : v2)
            if (x.first != y.first.substr(2, 3))
            {
                temp.push_back(x);
                val += x.second;
                if (temp.size() == 4)
                    break;
            }

        if (temp.size() != 4 && val + y.second < n)
            maxn = max(maxn, (long long)y.second * 4000 + val * 500 + 300000);

        if (temp.size() == 1)
            maxn = max({maxn, (long long)y.second * 4000 + temp[0].second * 500, (long long)y.second * 4000 + 300000});
        else if (temp.size() == 4)
            maxn = max({maxn, (long long)y.second * 4000 + 500 * (temp[0].second + temp[1].second + temp[2].second) + 300000});
        else
        {
            long long d = 0;
            for (int i = 0; i < temp.size(); i++)
            {
                maxn = max(maxn, (long long)y.second * 4000 + d * 500 + 300000);
                d = (long long)d + temp[i].second;
                maxn = max(maxn, (long long)y.second * 4000 + d * 500);
                // if (d+y.second < n) maxn=max(maxn,(long long)y.second*4000 + d*500 + 300000);
            }
        }
        // cout<<maxn<<"\n";
    }
    cout << maxn;
}

File: ICPC/The 2020 ICPC Asia Yokohama Regional Contest/B.py
from collections import defaultdict 

def tri(n,data):
    count2 = defaultdict(lambda:0)
    count4 = defaultdict(lambda:0)
    count6 = defaultdict(lambda:0)
    for s in data:
        count4[s[2:]]+=1
        count2[s[4:]]+=1
        count6[s]+=1



    for num in count6:
        count6[num]*=300000

    for num in count4:
        count4[num]*=4000

    for num in count2:
        count2[num]*=500


    data6=defaultdict(lambda:0)
    for num in count6:
        data6[num]=(count6[num]-count4[num[2:]]-count2[num[4:]])

    res=0
    maxVal = max(data6.values())
    for num in data6:
        if data6[num]==maxVal:
            count4[num[2:]]=0
            count2[num[4:]]=0
            res+=count6[num]
            break
    data4=defaultdict(lambda:0)
    for num in count4:
        data4[num]=(count4[num]-count2[num[2:]])

    maxVal = max(data4.values())
    for num in data4:
        if data4[num]==maxVal:
            count2[num[2:]]=0
            res+=count4[num]
            break

    data2 = list(count2.values())
    data2.sort(reverse=True)
    res+=sum(data2[:3])
    return (res)






def nhat(n,data):
    d4 = {}
    d2 = {}
    
    for s in data:
        t1 = s[2:7]
        t2 = s[4:7]
        d4[t1] = d4.get(t1, 0) + 1
        d2[t2] = d2.get(t2, 0) + 1

    v1 = [(key, value) for key, value in d4.items()]
    v2 = [(key, value) for key, value in d2.items()]

    def cmp(x, y):
        return x[1] > y[1]

    v2.sort(key=lambda x: x[1], reverse=True)
    
    temp = []
    maxn = 0

    for x in v2:
        temp.append(x)
        if len(temp) == 4:
            break

    if len(temp) == 1:
        maxn = max(maxn, temp[0][1] * 500)
        maxn = max(maxn, 300000)
    elif len(temp) == 4:
        maxn = max(maxn, 500 * (temp[0][1] + temp[1][1] + temp[2][1]) + 300000)
    else:
        d = 0
        for i in range(len(temp)):
            maxn = max(maxn, d * 500 + 300000)
            d += temp[i][1]
            maxn = max(maxn, d * 500)


    for y in v1:
        temp = []
        maxn = max(maxn, y[1] * 4000 + 300000 * (n != y[1]))
        val = 0
        
        for x in v2:
            if x[0] != y[0][2:5]:
                temp.append(x)
                val += x[1]
                if len(temp) == 4:
                    break
        
        if len(temp) != 4 and val + y[1] < n:
            maxn = max(maxn, y[1] * 4000 + val * 500 + 300000)
        
        if len(temp) == 1:
            maxn = max(maxn, y[1] * 4000 + temp[0][1] * 500)
            maxn = max(maxn, y[1] * 4000 + 300000)
        elif len(temp) == 4:
            maxn = max(maxn, y[1] * 4000 + 500 * (temp[0][1] + temp[1][1] + temp[2][1]) + 300000)
        else:
            d = 0
            for i in range(len(temp)):
                maxn = max(maxn, y[1] * 4000 + d * 500 + 300000)
                d += temp[i][1]
                maxn = max(maxn, y[1] * 4000 + d * 500)
    return (maxn)

import random

for t in range(100000):
    n = random.randint(1,10)
    data=[]
    for i in range(n):
        temps=""
        for i in range(6):
            temp=random.randint(0,9)
            temps+=str(temp)
        data.append(temps)

    
    n = len(set(data))
    data = list(set(data))

    # print(tri(n,data), nhat(n,data))
    if tri(n,data)>nhat(n,data):
        print(n)
        print(data)
        print(tri(n,data))
        print(nhat(n,data))
        print("NHAT SAI")
    elif tri(n,data)< nhat(n,data):
        print(n)
        print(data)
        print(tri(n,data))
        print(nhat(n,data))
        print("TRI SAI")

